### DISCUSSION TITLE
Publish to MS Store

### DISCUSSION THREAD

**Discussion by imba-tjd** [microsoft/openjdk#13] [Q&A] [answered]
Just like python

![图片](https://user-images.githubusercontent.com/24759802/113886643-6e66a580-97f3-11eb-80b1-b647bd511ccb.png)

**Comment by karianna** [comment]
Thanks for the feedback, we'll discuss it internally (it may take a while for a decision to come out the other side, but we'll do our best to update you).

**Comment by brunoborges** [comment, marked as answer]
Hi @imba-tjd,

Have you tried `winget`? It is the new package manager for Windows 10. After [installing it](https://docs.microsoft.com/en-us/windows/package-manager/winget/#install-winget), you can grab the Microsoft Build of OpenJDK like this:

![image](https://user-images.githubusercontent.com/129743/113947196-f3839780-97be-11eb-8090-8243d400b760.png)

**Reply by imba-tjd** [reply]
I know winget. I didn't use it. I think it's more like a downloader.

Store app can update automatically, and can be uninstalled easily. It's also convenient for beginners who don't konw how to use ~~CLI and setup PATH~~. The msi installer is already easy to setup and can add path.

**Reply by brunoborges** [reply]
> I know winget. I didn't use it. I think it's more like a downloader.
> 
> Store app can update automatically, and can be uninstalled easily. It's also convenient for beginners who don't konw how to use CLI and setup PATH.

Let's say Java 11 is installed. Would you expect the update to move to Java 12 after 6 months?

Or should the update only apply to minor versions (e.g. 11.0.current to 11.0.next, every quarter)?

**Reply by imba-tjd** [reply]
I don't have experience in java :joy:.
For py, they update minor version. Py has alias like `python3.8` so that it's ok to install multiple version.

Java is known as good compatibility and is very cautious in adding features. I personally think it's ok to upgrade major version. Just keep the LTS version to not upgrade major version.

**Reply by Moodjbow** [reply]
Dear all,
Can we please sort the wheat from the chaff?
Store apps, similar to new Win10 OS, are upgraded twice per anno approx. And I mean upgrade, not update.
If OpenJDK upgrade were backwards compatible with previous major versions, all good. But it's not. Then think of the millions of legacy business applications still running on V8. That's where Oracle consigned the torch.
Long-term update commitment to major versions sounds safer both for business and developers.
In short, if Microsoft can provide seamless updates per major version for business apps users, that'd be great!

**Reply by danielniccoli** [reply]
> > I know winget. I didn't use it. I think it's more like a downloader.
> > Store app can update automatically, and can be uninstalled easily. It's also convenient for beginners who don't konw how to use CLI and setup PATH.
> 
> Let's say Java 11 is installed. Would you expect the update to move to Java 12 after 6 months?
> 
> Or should the update only apply to minor versions (e.g. 11.0.current to 11.0.next, every quarter)?

You can handle this like you do with Ubuntu/WSL. Create LTS apps that stay in the major version and create a Java LATEST that is ... always the latest :)

**Reply by pratikpc** [reply]
> Let's say Java 11 is installed. Would you expect the update to move to Java 12 after 6 months?

@brunoborges  you could have separate releases for versions and a Latest Release

That would solve this issue

**Reply by Moodjbow** [reply]
In essence, all agree here that major versions should have their own life-cycle management as a store app. Dear @karianna, did you have a chance to discuss this internally and come to a conclusion?

**Reply by karianna** [reply]
At this stage we're leaning towards best practice with modern Java applications.  That is, they should be deployed as native applications with the parts of the Java runtime and the application code bundled in (i.e. With jlink and jpackage).  Installing a full JDK/JRE as a standalone package for an end user is something that the whole Java industry wants to get away from.

**Reply by treestryder** [reply]
I love the dream. But still have to re-package every update forJava devolopers and some apps that have yet to read the memo to implement the dream.

**Reply by Moodjbow** [reply]
Courageous, hats off! Roughly 10% of all java apps come packaged with the runtime and other dependencies, so there's only 90% left to re-educate. And considering java runtimes get security patches monthly, the java developers need to be very disciplined to re-package their products monthly even if they haven't changed a line of their code. Oh, well, we live in virtual reality, so why not dream a bit more ;)

**Reply by karianna** [reply]
We hear your pain of the Chicken and Egg problem! Will definitely keep discussing here and internally!

**Reply by danielniccoli** [reply]
> At this stage we're leaning towards best practice with modern Java applications. That is, they should be deployed as native applications with the parts of the Java runtime and the application code bundled in (i.e. With jlink and jpackage). Installing a full JDK/JRE as a standalone package for an end user is something that the whole Java industry wants to get away from.

That comment is off-topic to the request, to have OpenJDK available in the MS Store with a long-term support release and short-term support release that keeps itself at the most current version.

**Reply by danielniccoli** [reply]
> We hear your pain of the Chicken and Egg problem! Will definitely keep discussing here and internally!

Will this discussion reach any conclusion? I don't see much progress. Can you please share with us the status of your internal discussion?

**Reply by karianna** [reply]
> > At this stage we're leaning towards best practice with modern Java applications. That is, they should be deployed as native applications with the parts of the Java runtime and the application code bundled in (i.e. With jlink and jpackage). Installing a full JDK/JRE as a standalone package for an end user is something that the whole Java industry wants to get away from.
> 
> That comment is off-topic to the request, to have OpenJDK available in the MS Store with a long-term support release and short-term support release that keeps itself at the most current version.

It's actually tied together since it's a 'Do we distribute full JDKs (JREs) which is considered a legacy behaviour or not'.

**Reply by karianna** [reply]
> > We hear your pain of the Chicken and Egg problem! Will definitely keep discussing here and internally!
> 
> Will this discussion reach any conclusion? I don't see much progress. Can you please share with us the status of your internal discussion?

I'm afraid this is a deeply complex topic (it's not just a matter of technical implementation), we'll certainly update here as we hit new decision points but otherwise those discussions have to remain internal with no guaranteed ETA.  I appreciate this is frustrating for all folks who are following this!

**Reply by danielniccoli** [reply]
> > > At this stage we're leaning towards best practice with modern Java applications. That is, they should be deployed as native applications with the parts of the Java runtime and the application code bundled in (i.e. With jlink and jpackage). Installing a full JDK/JRE as a standalone package for an end user is something that the whole Java industry wants to get away from.
> > 
> > 
> > That comment is off-topic to the request, to have OpenJDK available in the MS Store with a long-term support release and short-term support release that keeps itself at the most current version.
> 
> It's actually tied together since it's a 'Do we distribute full JDKs (JREs) which is considered a legacy behaviour or not'.

I believe the question should be a simple 'Do we distribute full JDKs (JREs)'.

What would you achieve if you decide not to publish the JDK and JRE to the MS Store? Nothing, because it wouldn't change the status quo. If you believe that bundling it should be the way to go, then "doing nothing" will not change anyone's mind about it. Because not publishing it to the store is basically just doing nothing.

On the other hand, what would happen, if you would publish them to the MS Store? You'd make it much easier for the consumers to get the kind of Java apps running, that would otherwise require you to download the JRE. With the added upside that their Java version would stay up to date, and secure, and vulnerabilities would be patched automatically - as opposed to never.

**Reply by brunoborges** [reply]
Thanks @danielniccoli, for these amazing insights. We are collecting as much feedback as possible and working with the .NET team, who has gone through these same problems/questions.

For now, I will say we do not have any plans to do install/update through the MS Store, and we don't have any sight of when we might come to any conclusions as we prioritize other items in our backlogs.

**Reply by PiprTuff** [reply]
Hello @brunoborges, any update on this till now? If you need opinions from Java enthusiasts and developers who have worked with from old JDKs and JREs to 
modern JPackage, then here is mine:

As Microsoft Store is now a great place to get apps and softwares easily, with auto updates, it is also a right place for developer tools to be published to. I see that there is Python available in the store, with all its latest minor versions as separate packages. Python is needed both during the development time and runtime for apps, so both developers and end users need it. But for Java (i.e. modern Java), only the developers need JDKs to develop apps, and then a smaller image consisting of only the needed modules of the JDK with all the dependencies are built-into the apps as runtimes using the JPackage tool, so the end users do not need to install Java themselves. So we see that JDKs should be available in the store, for developers, and there is no need for JREs.

Now come to versions. OpenJDK 8, 11, and 17 has been declared to be LTS versions, meaning they will receive bug fixes and improvements updates for long period. Among them, OpenJDK 8's support period has end. Also, the latest version of the OpenJDK in a time (18 as of now), be it STS (short term supported), has all the latest bug fixes and improvements with new features added. So I think, at least the LTS releases and the latest release (be it STS) should be published into the store.

So for now, there should be at least 3 packages:
1. OpenJDK 11 (LTS)
2. OpenJDK 17 (LTS)
3. OpenJDK Latest (currently 18 STS)
4. (and beyond) If you wish, you can also add OpenJDK 8 (which Oracle said to support until 2030, under commercial plans), and one or two recent versions, like OpenJDK 16.

Now come to updates. Although Java is backward compatible, new bugs are always arriving with new major releases. Most of the Java developers prefer to stick to a LTS version, so that they don't need to change their codebases regularly for each major JDK update. They simply updates to next minor releases which come with bug fixes and improvements for the existing version. Very few prefer to be updated with the latest Java release.

So, in the store, OpenJDK 11 should get updates for its next minor releases (like 11.0.2+7 to 11.0.13+8), and same for OpenJDK 17. But if anyone installs the OpenJDK Latest package, it should be updated to any next release found of the latest Java version, be it major or minor (like 18.0.5+3 to 18.0.9+13 or better 19.0.1+2, if found).

**Comment by pauldaustin** [comment]
I would also like to see it published to the Microsoft Store. I have a client who is using Microsoft Endpoint Manager (Intune) to manage applications on workstations. For that publishing Microsoft Store apps is the easiest as you get auto updates.

For me I would like to see Microsoft OpenJDK 11, Microsoft OpenJDK 16 which would be tied to major versions. And also allow multiple different installs.

Install path would be something like %PROGRAM_FILES%\Microsoft\OpenJDK\11 (no patch number so that you aren't constantly updating paths).

**Reply by brunoborges** [reply]
hey @pauldaustin, I'm not familiar with MEM, but I'm curious if it doesn't support already MSI packages from the central/admin dashboard. Do you know?

**Reply by treestryder** [reply]
Sure, it is possible to download an MSI, repackage (essentially zip) it, and push it to thousands of PCs via MEM. But then I have added managing that application to my workload, indefinitely. Somehow, I have to monitor each application for security updates, carefully deploy those updates, AND I am responsible when those updates break something. The Microsoft Store moves that work and responsibility to the software authors.

Packaging the software as an MSIX or APPX package also has security, reliability and management benefits.

**Comment by KalleOlaviNiemitalo** [comment]
Windows Terminal has got a few complaints about Microsoft Store updating it and terminating its processes without user consent: <https://github.com/microsoft/terminal/issues/3915>, <https://github.com/microsoft/terminal/issues/6726>. I suspect that, if Microsoft Store terminates users' Java apps when it updates OpenJDK, that will elicit similar complaints.

**Reply by treestryder** [reply]
I wonder if MSIX package features could be taken advantage to "ensure that only the essential changed bits of the app are downloaded to update an existing Windows app."
https://docs.microsoft.com/en-us/windows/msix/app-package-updates
